Multipart copy now uses a thread pool (resolves #859)

JakeNarkizian opened the pull request:
Copies within the same region should now be much faster.

JakeNarkizian commented:
@hannes-ucsc

hannes-ucsc commented:
Commit message looks bjørked (a term coined by @jvivian).

hannes-ucsc commented on src/toil/jobStores/aws/jobStore.py:
```diff
@@ -1047,3 +1011,75 @@ def toItem(self):
+    s3 = boto.connect_s3()
+    partSize = defaultPartSize
+    totalSize = srcKey.size
+    totalParts = (totalSize / partSize) + 1 if totalSize % partSize else (totalSize / partSize)
```
`totalParts = (totalSize + partSize - 1) / partSize`

hannes-ucsc replied:
Besides, all the parentheses in your expression are redundant. Unless the operators are obscure (and their precedence not obvious), I prefer omitting parentheses. Everyone knows that / and % bind stronger than + which binds stronger than the conditional.

hannes-ucsc commented on src/toil/jobStores/aws/jobStore.py:
```diff
@@ -1047,3 +1011,75 @@ def toItem(self):
+        else:
+            try:
+                startByte = partIndex * partSize
+                startByte = startByte if startByte <= totalSize else totalSize - ((partIndex - 1) * partSize)
```
Explain. How can `partIndex * partSize` ever be `>= totalSize`?

JakeNarkizian replied:
Assuming ```n``` parts, the nth part is necessarily smaller than ```partSize``` if and only if ```totalSize % partSize != 0```.

Assuming ```totalSize % partSize != 0```. Every part except the nth and final part must be of size ```partSize```, thus  ```totalSize = (n-1) * partSize + sizeof(n)```. 
```n * partSize =  (n-1) * partSize + partSize ```
since ```partSize > sizeof(n) ``` by the assumed case,
 ```n * partSize > totalSize ``` must be true in this case. 

Assuming ```totalSize % partSize == 0``` we have that ```n * partSize = totalSize ``` since ```sizeof(n)``` is necessarily ```partSize```.

Therefore ```partIndex * partSize >= totalSize``` must be true when ```partIndex = n```.

hannes-ucsc replied:
Sorry, but no. With `partSize=2` and `totalSize=3` you get `n==2` and `partIndex in [0,1]`. When partIndex is 1 (the last part) you get startByte == partSize which is 2 and less than 3.

hannes-ucsc replied:
And partIndex should never be `== n`, that's probably where your mistake lies. The biggest value it gets should be n - 1.

JakeNarkizian replied:
You're right what I said is irrelevant, I overlooked the fact that ```0 <= partIndex < n```.

hannes-ucsc commented on src/toil/jobStores/aws/jobStore.py:
```diff
@@ -1047,3 +1011,75 @@ def toItem(self):
+    partSize = defaultPartSize
+    totalSize = srcKey.size
+    totalParts = (totalSize / partSize) + 1 if totalSize % partSize else (totalSize / partSize)
+    exceptionList = []
```
`exceptions`

hannes-ucsc commented on src/toil/jobStores/aws/jobStore.py:
```diff
@@ -1047,3 +1011,75 @@ def toItem(self):
+
+            if exceptionList:
+                i = 0
+                for exceptionInfo in exceptionList:
```
`in exceptionList[0:4]`

hannes-ucsc replied:
But see above.

hannes-ucsc replied:
Once you switch `exceptionList`/`exceptions` to a ordered dict, simply log every entry. You may have to switch on the type of the value (traceback string or exception) and do sth special in each case but I don't think so.

hannes-ucsc commented on src/toil/jobStores/aws/jobStore.py:
```diff
@@ -1047,3 +1011,75 @@ def toItem(self):
+                log.debug('Copied part %d of %d' % (partIndex, totalParts))
+
+            except:
+                exceptionList.append({'partIndex': partIndex, 'tb': traceback.format_exc()})
```
Change of plan. Since we only log the first four backtraces, use `format_exc` only for the first four. For the rest, use just the exception object.

Also make `exceptionList` an OrderedDict and use the `partIndex` as a key and the traceback/exception as the value.

Rename `exceptionList` to `exceptions`. I don't like encoding the type in variable names.

hannes-ucsc commented on src/toil/jobStores/aws/jobStore.py:
```diff
@@ -1047,3 +1011,75 @@ def toItem(self):
+                    else:
+                        log.info("Part %d of multipart copy failed with the following exception:\n%s" %
+                                 (exceptionInfo['partIndex'], exceptionInfo['tb']))
+                        i += 1
```
Don't you also need to raise an exception as the last element in the if body, such that cancel_upload is triggered?

hannes-ucsc replied:
I mean the body of the `if exceptionList`

benedictpaten commented:
Looks good. Is this parallel upload functionality tested during the job store test? I assume it is - but I'm wondering if it is a sufficiently stressful test.

JakeNarkizian commented:
@benedictpaten Yes, the file staging tests (which are generated at runtime) covers the copy functionality when importing from an AWS bucket that is external to the job store.  

These tests as they stand employ thread pool sizes of 1 and 2. Since the thread pool size is dynamically set ([see](https://github.com/BD2KGenomics/toil/blob/issues/859-multipart-copy-threadpool/src/toil/jobStores/aws/jobStore.py#L1068)), they do not test any number of threads greater than two. What do you think of adding a separate multipart copy test that employs a thread count of 16 and 128?

JakeNarkizian commented:
tests failed because there were not enough buckets or domains. The problem should be fixed now. 

Jenkins, test this please.

JakeNarkizian commented:
Jenkins, test this please.

JakeNarkizian commented:
Jenkins, test this please.

hannes-ucsc commented on src/toil/jobStores/aws/jobStore.py:
```diff
@@ -1048,3 +1014,77 @@ def toItem(self):
+                    log.debug("Part %d of multipart copy failed with the following exception:\n%s" %
+                              (partIndex, exception))
+
+                raise RuntimeError()
```
Add message: `"One or more parts failed to copy. See log for details."`

hannes-ucsc commented on src/toil/jobStores/aws/jobStore.py:
```diff
@@ -1048,3 +1014,77 @@ def toItem(self):
+    totalSize = srcKey.size
+    totalParts = (totalSize + partSize - 1) / partSize
+    exceptions = collections.OrderedDict()
+    upload = boto.connect_s3().get_bucket(dstBucketName).initiate_multipart_upload(dstKeyName,
```
Can you break this up into multiple statements. It looks awkward.

hannes-ucsc commented on src/toil/test/src/multipartTransferTest.py:
```diff
@@ -0,0 +1,106 @@
+
+
+@contextmanager
+def openAWS(keySize=None):
```
Bad name. This is S3 not AWS.

hannes-ucsc commented on src/toil/test/src/multipartTransferTest.py:
```diff
@@ -0,0 +1,106 @@
+
+        try:
+            with open('/dev/urandom', 'rb') as f:
+                with ThreadPoolExecutor(max_workers=min(cpu_count() * 16, totalParts, 128)) as executor:
```
Remove use of ThreadPoolExecutor here. It's not worth it for a short test. Simplify, simplify, simplify.

JakeNarkizian replied:
Are you sure? Uploading the random bytes to set up the test will take a significantly longer time.

hannes-ucsc replied:
> significantly more time

Can you back that up with measurements?

And yes, I am sure. I don't want to amass all this code for the sake of testing one teeny  component of of Toil. Keep it simple Jake, please! When we identify a performance bottleneck, we can revisit. Right now it is more important to make the code easy to maintain.

JakeNarkizian replied:
Ok, I see your point. 

Tangentially, if you would like I can benchmark it, but informally yesterday uploading ```128 * partSize``` took about 5 minutes when multithreaded. I terminated the single thread run after it was still going for about 25 minutes. 

Also, assuming 128 is omitted, if ```16 * partSize``` is significantly slower that a multithreaded approach would that sway you to keep it? I can calculate the different speeds for this as well if you'd like.

hannes-ucsc replied:
It will save a few tens of seconds, sure. But it uses minutes of my time since I have to review it. Faster isn't always better. Simpler almost always is.

hannes-ucsc replied:
When you reduce the part size to 5MiB and the pool size to 16 this does not need to be a multipart copy. Remove.

hannes-ucsc commented on src/toil/test/src/multipartTransferTest.py:
```diff
@@ -0,0 +1,106 @@
+    s3 = boto.connect_s3()
+    bucket = s3.create_bucket('multipart-transfer-test-%s' % uuid.uuid4())
+    keyName = uuid.uuid4()
+    if keySize is not None:
```
invert condition so you can have the `yield bucket` at the top and not dangling at the bottom.

hannes-ucsc commented on src/toil/test/src/multipartTransferTest.py:
```diff
@@ -0,0 +1,106 @@
+                    assert srcKey.get_contents_as_string() == dstBucket.get_key("test").get_contents_as_string()
+
+        make_tests(_testMultipartCopy, targetClass=AWSMultipartCopyTest,
+                   threadPoolSize={str(x): x for x in (1, 2, 16, 128)})
```
Eliminate 128.

JakeNarkizian replied:
Isn't it worth covering as a corner case?

hannes-ucsc replied:
Why would 128 fail if 16 suceeded?

JakeNarkizian replied:
That actually happened yesterday see [this](http://jenkins.cgcloud.info/job/toil-pull-requests/1314/).

JakeNarkizian replied:
albeit it happened in the ```openAWS``` set up method. However the skeleton of that method is identical to that of the copy.

hannes-ucsc replied:
Seems that failure could have happened with 2 or 16. What specifically about the 128 case caused that failure?

hannes-ucsc commented:
Re: more stressful test: Yes, but remove the 128 case. Toil tests are already taking much too long severely limiting our PR RTT.

Also @jake why do you think the existing tests won't use a thread pool larger than two?

JakeNarkizian commented:
@hannes-ucsc The import export tests import from some job store storage mechanism to some job store for all currently supported job stores. Recall each of these pairings is tested with a set of sizes two of which are ``` partSize``` and ```partSize + 1```.  Thus two of these tests will import from S3 to the AWS job store keys of the two sizes mentioned above. The first of these tests will be a single part and will use only one thread. The second will have two parts and thus two threads.

JakeNarkizian commented:
Jenkins, test this please.

JakeNarkizian commented:
@hannes-ucsc not sure why it says failed... the Jenkins build does not show that any tests have failed. Otherwise ready.

hannes-ucsc commented:
@JakeNarkizian, look at the end of the build's console output.

tetron commented:
```
mesosbox@ip-172-31-22-175:~/scale-test-wf$ cwltoil --logDebug --batchSystem=mesos --mesosMaster=mesos-master:5050 --jobStore=aws:us-west-2:cwltoil-test-pa2 main-bwa_freebayes_HG01977.cwl main-bwa_freebayes_HG01977-samples.json
ip-172-31-22-175: 2016-06-01 17:41:40,951 INFO: toil.lib.bioio: Logging set at level: DEBUG
ip-172-31-22-175: 2016-06-01 17:42:15,707 INFO: toil.lib.bioio: Logging set at level: DEBUG
Traceback (most recent call last):
  File "/usr/local/bin/cwltoil", line 9, in <module>
    load_entry_point('toil==3.2.0a2', 'console_scripts', 'cwltoil')()
  File "/usr/local/lib/python2.7/dist-packages/toil/cwl/cwltoil.py", line 577, in main
    with Toil(options) as toil:
  File "/usr/local/lib/python2.7/dist-packages/toil/common.py", line 487, in __enter__
    config=None if self.config.restart else self.config)
  File "/usr/local/lib/python2.7/dist-packages/toil/common.py", line 602, in loadOrCreateJobStore
    from toil.jobStores.aws.jobStore import AWSJobStore
  File "/usr/local/lib/python2.7/dist-packages/toil/jobStores/aws/jobStore.py", line 46, in <module>
    from concurrent.futures import ThreadPoolExecutor
ImportError: No module named concurrent.futures
```

I think you're missing a dependency in setup.py in order to run on Python 2.7

tetron commented:
Oops, it failed:

```
mesosbox@ip-172-31-22-175:~/scale-test-wf$ toil clean aws:us-west-2:cwltoil-test-pa2 && cwltoil --logDebug --batchSystem=mesos --mesosMaster=mesos-master:5050 --jobStore=aws:us-west-2:cwltoil-test-pa2 main-bwa_freebayes_HG01977.cwl main-bwa_freebayes_HG01977-samples.json
ip-172-31-22-175: 2016-06-01 17:47:14,366 INFO: toil.lib.bioio: Logging set at level: INFO
ip-172-31-22-175: 2016-06-01 17:47:14,366 INFO: toil.utils.toilClean: Parsed arguments
ip-172-31-22-175: 2016-06-01 17:47:14,366 INFO: toil.utils.toilClean: Checking if we have files for toil
ip-172-31-22-175: 2016-06-01 17:47:14,507 INFO: toil.utils.toilClean: The specified JobStore does not exist, it may have already been deleted
ip-172-31-22-175: 2016-06-01 17:47:14,824 INFO: toil.lib.bioio: Logging set at level: DEBUG
ip-172-31-22-175: 2016-06-01 17:47:50,929 INFO: toil.lib.bioio: Logging set at level: DEBUG
ip-172-31-22-175: 2016-06-01 17:47:51,002 DEBUG: toil.jobStores.aws.jobStore: Instantiating <class 'toil.jobStores.aws.jobStore.AWSJobStore'> for region us-west-2 and name prefix 'cwltoil-test-pa2'
ip-172-31-22-175: 2016-06-01 17:47:54,577 INFO: toil.jobStores.abstractJobStore: The workflow ID is: 'ca4b8a47-467d-4bac-b393-3b20b25e8d8b'
ip-172-31-22-175: 2016-06-01 17:47:59,706 DEBUG: toil.jobStores.aws.jobStore: Wrote FileInfo(fileID='a77581f6-ad82-5142-acb7-5a6b01eab3fb', ownerID='891f7db6-e4d9-4221-a58e-ab6cc4395f94', encrypted=False, version='', previousVersion='', content='\x80\x02ctoil.co...ryCountq:K\x00ub.', _numContentChunks=2) bound to <toil.jobStores.aws.jobStore.AWSJobStore object at 0x7f4531bc8510> for shared file 'config.pickle'.
ip-172-31-22-175: 2016-06-01 17:47:59,802 INFO: toil.jobStores.abstractJobStore: Unable to import 'toil.jobStores.azureJobStore'
ip-172-31-22-175: 2016-06-01 17:48:00,329 INFO: toil.jobStores.aws.jobStore: Initiated multipart copy from 's3://1000genomes/phase3/data/HG01977/sequence_read/SRR360135_1.filt.fastq.gz' to 's3://cwltoil-test-pa2--files/f24fabaa-e9bb-4f9c-b527-bc16d1ff9229'.
ip-172-31-22-175: 2016-06-01 17:48:50,883 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 0 of 258 of 
ip-172-31-22-175: 2016-06-01 17:48:50,942 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 12 of 258 of 
ip-172-31-22-175: 2016-06-01 17:48:50,970 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 27 of 258 of 
ip-172-31-22-175: 2016-06-01 17:48:50,992 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 26 of 258 of 
ip-172-31-22-175: 2016-06-01 17:48:50,995 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 22 of 258 of 
ip-172-31-22-175: 2016-06-01 17:48:51,069 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 8 of 258 of 
ip-172-31-22-175: 2016-06-01 17:48:51,131 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 29 of 258 of 
ip-172-31-22-175: 2016-06-01 17:48:51,409 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 16 of 258 of 
ip-172-31-22-175: 2016-06-01 17:48:51,603 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 19 of 258 of 
ip-172-31-22-175: 2016-06-01 17:48:51,983 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 25 of 258 of 
ip-172-31-22-175: 2016-06-01 17:48:52,334 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 5 of 258 of 
ip-172-31-22-175: 2016-06-01 17:48:52,540 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 2 of 258 of 
ip-172-31-22-175: 2016-06-01 17:48:52,705 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 9 of 258 of 
ip-172-31-22-175: 2016-06-01 17:48:52,903 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 31 of 258 of 
ip-172-31-22-175: 2016-06-01 17:48:52,904 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 18 of 258 of 
ip-172-31-22-175: 2016-06-01 17:48:52,921 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 17 of 258 of 
ip-172-31-22-175: 2016-06-01 17:48:53,185 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 6 of 258 of 
ip-172-31-22-175: 2016-06-01 17:48:53,410 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 21 of 258 of 
ip-172-31-22-175: 2016-06-01 17:48:53,477 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 30 of 258 of 
ip-172-31-22-175: 2016-06-01 17:48:53,986 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 7 of 258 of 
ip-172-31-22-175: 2016-06-01 17:48:54,199 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 23 of 258 of 
ip-172-31-22-175: 2016-06-01 17:48:54,809 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 15 of 258 of 
ip-172-31-22-175: 2016-06-01 17:48:54,836 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 3 of 258 of 
ip-172-31-22-175: 2016-06-01 17:48:55,556 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 11 of 258 of 
ip-172-31-22-175: 2016-06-01 17:48:56,619 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 20 of 258 of 
ip-172-31-22-175: 2016-06-01 17:48:56,748 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 1 of 258 of 
ip-172-31-22-175: 2016-06-01 17:48:57,833 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 28 of 258 of 
ip-172-31-22-175: 2016-06-01 17:49:13,454 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 24 of 258 of 
ip-172-31-22-175: 2016-06-01 17:49:16,140 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 13 of 258 of 
ip-172-31-22-175: 2016-06-01 17:49:17,234 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 10 of 258 of 
ip-172-31-22-175: 2016-06-01 17:49:24,206 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 4 of 258 of 
ip-172-31-22-175: 2016-06-01 17:49:28,217 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 14 of 258 of 
ip-172-31-22-175: 2016-06-01 17:49:41,416 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 36 of 258 of 
ip-172-31-22-175: 2016-06-01 17:49:41,427 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 34 of 258 of 
ip-172-31-22-175: 2016-06-01 17:49:42,373 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 38 of 258 of 
ip-172-31-22-175: 2016-06-01 17:49:44,063 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 41 of 258 of 
ip-172-31-22-175: 2016-06-01 17:49:44,529 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 42 of 258 of 
ip-172-31-22-175: 2016-06-01 17:49:44,762 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 33 of 258 of 
ip-172-31-22-175: 2016-06-01 17:49:44,925 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 49 of 258 of 
ip-172-31-22-175: 2016-06-01 17:49:45,017 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 39 of 258 of 
ip-172-31-22-175: 2016-06-01 17:49:45,832 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 47 of 258 of 
ip-172-31-22-175: 2016-06-01 17:49:46,509 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 45 of 258 of 
ip-172-31-22-175: 2016-06-01 17:49:46,539 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 55 of 258 of 
ip-172-31-22-175: 2016-06-01 17:49:46,692 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 51 of 258 of 
ip-172-31-22-175: 2016-06-01 17:49:46,701 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 32 of 258 of 
ip-172-31-22-175: 2016-06-01 17:49:46,763 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 44 of 258 of 
ip-172-31-22-175: 2016-06-01 17:49:46,772 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 52 of 258 of 
ip-172-31-22-175: 2016-06-01 17:49:46,929 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 50 of 258 of 
ip-172-31-22-175: 2016-06-01 17:49:47,416 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 53 of 258 of 
ip-172-31-22-175: 2016-06-01 17:49:48,529 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 57 of 258 of 
ip-172-31-22-175: 2016-06-01 17:49:48,820 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 58 of 258 of 
ip-172-31-22-175: 2016-06-01 17:49:49,150 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 40 of 258 of 
ip-172-31-22-175: 2016-06-01 17:49:49,369 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 54 of 258 of 
ip-172-31-22-175: 2016-06-01 17:49:49,865 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 46 of 258 of 
ip-172-31-22-175: 2016-06-01 17:49:51,077 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 35 of 258 of 
ip-172-31-22-175: 2016-06-01 17:49:58,617 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 43 of 258 of 
ip-172-31-22-175: 2016-06-01 17:50:00,032 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 37 of 258 of 
ip-172-31-22-175: 2016-06-01 17:50:12,414 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 60 of 258 of 
ip-172-31-22-175: 2016-06-01 17:50:14,174 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 56 of 258 of 
ip-172-31-22-175: 2016-06-01 17:50:16,173 DEBUG: toil.jobStores.aws.jobStore: Copy failed.
ip-172-31-22-175: 2016-06-01 17:50:28,634 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 48 of 258 of 
ip-172-31-22-175: 2016-06-01 17:50:31,391 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 59 of 258 of 
ip-172-31-22-175: 2016-06-01 17:50:31,940 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 65 of 258 of 
ip-172-31-22-175: 2016-06-01 17:50:31,958 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 64 of 258 of 
ip-172-31-22-175: 2016-06-01 17:50:32,905 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 66 of 258 of 
ip-172-31-22-175: 2016-06-01 17:50:35,294 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 69 of 258 of 
ip-172-31-22-175: 2016-06-01 17:50:36,186 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 68 of 258 of 
ip-172-31-22-175: 2016-06-01 17:50:36,912 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 67 of 258 of 
ip-172-31-22-175: 2016-06-01 17:50:37,126 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 71 of 258 of 
ip-172-31-22-175: 2016-06-01 17:50:37,282 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 76 of 258 of 
ip-172-31-22-175: 2016-06-01 17:50:37,476 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 70 of 258 of 
ip-172-31-22-175: 2016-06-01 17:50:37,605 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 74 of 258 of 
ip-172-31-22-175: 2016-06-01 17:50:38,228 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 73 of 258 of 
ip-172-31-22-175: 2016-06-01 17:50:38,736 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 75 of 258 of 
ip-172-31-22-175: 2016-06-01 17:50:39,177 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 77 of 258 of 
ip-172-31-22-175: 2016-06-01 17:50:39,398 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 82 of 258 of 
ip-172-31-22-175: 2016-06-01 17:50:39,711 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 72 of 258 of 
ip-172-31-22-175: 2016-06-01 17:50:40,886 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 80 of 258 of 
ip-172-31-22-175: 2016-06-01 17:50:41,544 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 84 of 258 of 
ip-172-31-22-175: 2016-06-01 17:50:41,545 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 86 of 258 of 
ip-172-31-22-175: 2016-06-01 17:50:41,750 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 81 of 258 of 
ip-172-31-22-175: 2016-06-01 17:50:42,471 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 79 of 258 of 
ip-172-31-22-175: 2016-06-01 17:50:45,164 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 85 of 258 of 
ip-172-31-22-175: 2016-06-01 17:50:52,335 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 63 of 258 of 
ip-172-31-22-175: 2016-06-01 17:50:54,196 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 62 of 258 of 
ip-172-31-22-175: 2016-06-01 17:50:54,610 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 88 of 258 of 
ip-172-31-22-175: 2016-06-01 17:50:55,416 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 83 of 258 of 
ip-172-31-22-175: 2016-06-01 17:51:00,128 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 61 of 258 of 
ip-172-31-22-175: 2016-06-01 17:51:14,540 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 89 of 258 of 
ip-172-31-22-175: 2016-06-01 17:51:24,581 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 87 of 258 of 
ip-172-31-22-175: 2016-06-01 17:51:30,344 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 90 of 258 of 
ip-172-31-22-175: 2016-06-01 17:51:30,482 ERROR: root: Got exception 'unhashable type' while writing 's3://1000genomes/phase3/data/HG01977/sequence_read/SRR360135_1.filt.fastq.gz'
Traceback (most recent call last):
  File "/usr/local/bin/cwltoil", line 9, in <module>
    load_entry_point('toil==3.2.0a2', 'console_scripts', 'cwltoil')()
  File "/usr/local/lib/python2.7/dist-packages/toil/cwl/cwltoil.py", line 587, in main
    adjustFiles(builder.job, functools.partial(writeFile, toil.importFile, {}))
  File "/usr/local/lib/python2.7/dist-packages/cwltool/process.py", line 128, in adjustFiles
    adjustFiles(rec[d], op)
  File "/usr/local/lib/python2.7/dist-packages/cwltool/process.py", line 131, in adjustFiles
    adjustFiles(d, op)
  File "/usr/local/lib/python2.7/dist-packages/cwltool/process.py", line 131, in adjustFiles
    adjustFiles(d, op)
  File "/usr/local/lib/python2.7/dist-packages/cwltool/process.py", line 126, in adjustFiles
    rec["path"] = op(rec["path"])
  File "/usr/local/lib/python2.7/dist-packages/toil/cwl/cwltoil.py", line 162, in writeFile
    index[x] = (writeFunc(rp), os.path.basename(x))
  File "/usr/local/lib/python2.7/dist-packages/toil/common.py", line 678, in importFile
    return self._jobStore.importFile(srcUrl)
  File "/usr/local/lib/python2.7/dist-packages/toil/jobStores/abstractJobStore.py", line 265, in importFile
    return self._importFile(findJobStoreForUrl(url), url)
  File "/usr/local/lib/python2.7/dist-packages/toil/jobStores/aws/jobStore.py", line 246, in _importFile
    info.copyFrom(srcKey)
  File "/usr/local/lib/python2.7/dist-packages/toil/jobStores/aws/jobStore.py", line 814, in copyFrom
    headers=self._s3EncryptionHeaders()).version_id
  File "/usr/local/lib/python2.7/dist-packages/toil/jobStores/aws/jobStore.py", line 846, in _copyKey
    headers=headers)
  File "/usr/local/lib/python2.7/dist-packages/toil/jobStores/aws/jobStore.py", line 1073, in copyKeyMultipart
    for partIndex, tb in exceptions[0:4].items():
TypeError: unhashable type
```

JakeNarkizian commented:
```
toil.jobStores.aws.jobStore: Copy failed.
``` 
Is expected [see](https://github.com/BD2KGenomics/toil/blob/issues/859-multipart-copy-threadpool/src/toil/jobStores/aws/jobStore.py#L1054). But you bring up a good point @hannes-ucsc, I'll add the part number to the message.

hannes-ucsc commented:
@JakeNarkizian can you add a qualifying statement to the `Unable to import 'toil.jobStores.azureJobStore' log message. The message should explain why it might be expected to not being able to load the Azure job store.

hannes-ucsc commented on src/toil/jobStores/aws/jobStore.py:
```diff
@@ -1114,3 +1071,82 @@ def toItem(self):
         return self.jobStoreID, self.binaryToAttributes(cPickle.dumps(self))
+
+
+def copyKeyMultipart(srcKey, dstBucketName, dstKeyName, headers=None):
```
Don't move this around so I can see what actually changed.

hannes-ucsc replied:
IOW, move the function back to the old place, however wrong that place may be.

hannes-ucsc commented:
Can we have the words "import/export" in the commit title without going over the 72 character limit?

And like in the other PR, please limit line length to 100.

tetron commented:
This is working better, it was able to copy a number of files, but eventually failed with a new error:

```
ip-172-31-22-175: 2016-06-15 19:43:13,070 INFO: toil.lib.bioio: Logging set at level: DEBUG
ip-172-31-22-175: 2016-06-15 19:43:13,138 DEBUG: toil.jobStores.aws.jobStore: Instantiating <class 'toil.jobStores.aws.jobStore.AWSJobStore'> for region us-west-2 and name prefix 'cwltoil-test-pa2'
ip-172-31-22-175: 2016-06-15 19:43:16,822 INFO: toil.jobStores.abstractJobStore: The workflow ID is: '819a87ea-828d-4043-8172-e83030935ed7'
ip-172-31-22-175: 2016-06-15 19:43:17,179 DEBUG: toil.jobStores.aws.jobStore: Wrote FileInfo(fileID='a77581f6-ad82-5142-acb7-5a6b01eab3fb', ownerID='891f7db6-e4d9-4221-a58e-ab6cc4395f94', encrypted=False, version='', previousVersion='', content='\x80\x02ctoil.co...ryCountq:K\x00ub.', _numContentChunks=2) bound to <toil.jobStores.aws.jobStore.AWSJobStore object at 0x7f6df3560d10> for shared file 'config.pickle'.
ip-172-31-22-175: 2016-06-15 19:43:17,269 INFO: toil.jobStores.abstractJobStore: Unable to import 'toil.jobStores.azureJobStore'
...
copies a bunch of stuff successfully
...
ip-172-31-22-175: 2016-06-15 20:04:16,069 INFO: toil.jobStores.aws.jobStore: Initiated multipart copy from 's3://biodata/collections/GRCh37/snpeff/GRCh37.75/snpEffectPredictor.bin' to 's3://cwltoil-test-pa2--files/baf65e51-ad6e-4089-805e-652d89321bc1'.
ip-172-31-22-175: 2016-06-15 20:04:17,746 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 2 of 3 of 
ip-172-31-22-175: 2016-06-15 20:05:06,900 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 0 of 3 of 
ip-172-31-22-175: 2016-06-15 20:05:07,332 DEBUG: toil.jobStores.aws.jobStore: Successfully copied part 1 of 3 of 
ip-172-31-22-175: 2016-06-15 20:05:07,721 INFO: toil.jobStores.aws.jobStore: Completed copy from 's3://biodata/collections/GRCh37/snpeff/GRCh37.75/snpEffectPredictor.bin' to 's3://cwltoil-test-pa2--files/baf65e51-ad6e-4089-805e-652d89321bc1'.
ip-172-31-22-175: 2016-06-15 20:14:47,152 ERROR: root: Got exception 'Key does not exist.' while writing 's3://biodata/collections/GRCh37/snpeff/GRCh37.75/sequence.GL000191.1.bin'
Traceback (most recent call last):
  File "/usr/local/bin/cwltoil", line 9, in <module>
    load_entry_point('toil==3.2.0a2', 'console_scripts', 'cwltoil')()
  File "/usr/local/lib/python2.7/dist-packages/toil/cwl/cwltoil.py", line 592, in main
    adjustFiles(builder.job, functools.partial(writeFile, toil.importFile, {}))
  File "/usr/local/lib/python2.7/dist-packages/cwltool/process.py", line 128, in adjustFiles
    adjustFiles(rec[d], op)
  File "/usr/local/lib/python2.7/dist-packages/cwltool/process.py", line 131, in adjustFiles
    adjustFiles(d, op)
  File "/usr/local/lib/python2.7/dist-packages/cwltool/process.py", line 128, in adjustFiles
    adjustFiles(rec[d], op)
  File "/usr/local/lib/python2.7/dist-packages/cwltool/process.py", line 131, in adjustFiles
    adjustFiles(d, op)
  File "/usr/local/lib/python2.7/dist-packages/cwltool/process.py", line 126, in adjustFiles
    rec["path"] = op(rec["path"])
  File "/usr/local/lib/python2.7/dist-packages/toil/cwl/cwltoil.py", line 162, in writeFile
    index[x] = (writeFunc(rp), os.path.basename(x))
  File "/usr/local/lib/python2.7/dist-packages/toil/common.py", line 677, in importFile
    return self._jobStore.importFile(srcUrl)
  File "/usr/local/lib/python2.7/dist-packages/toil/jobStores/abstractJobStore.py", line 265, in importFile
    return self._importFile(findJobStoreForUrl(url), url)
  File "/usr/local/lib/python2.7/dist-packages/toil/jobStores/aws/jobStore.py", line 246, in _importFile
    srcBucket, srcKey = self._extractKeyInfoFromUrl(url, existing=True)
  File "/usr/local/lib/python2.7/dist-packages/toil/jobStores/aws/jobStore.py", line 288, in _extractKeyInfoFromUrl
    raise RuntimeError('Key does not exist.')
RuntimeError: Key does not exist.
```

hannes-ucsc commented:
I have to ask the obvious question: s3://biodata/collections/GRCh37/snpeff/GRCh37.75/sequence.GL000191.1.bin really does exist?

tetron commented:
That's a valid question.  I don't know, I will check.

hannes-ucsc commented:
@JakeNarkizian the exception should include the key and bucket in question.

Also, this branch doesn't have my import/export fixes that landed on master recently. Once the branch has been rebased, you might want to try again.

hannes-ucsc commented:
In fact, I did fix that one: #926.

tetron commented:
Great, I'll give it another try once the branch is rebased.

hannes-ucsc commented:
There are still some line notes that need to be addressed.

hannes-ucsc commented on src/toil/test/src/multipartTransferTest.py:
```diff
@@ -0,0 +1,108 @@
+    is yielded. If no keySize is given an empty bucket handle is yielded. The bucket and all
+    created keys are cleaned up automatically.
+
+    :param int keySize: Size of source key.
```
And where is the destination key?

JakeNarkizian replied:
I'm not sure what you mean here, could you elaborate?

hannes-ucsc replied:
Mentioning the word "source" implies that there also is a "destination" but in the context of this method, there is no destination. Worse, this method writes to the key, so calling it source is confusing.

hannes-ucsc commented on src/toil/test/src/multipartTransferTest.py:
```diff
@@ -0,0 +1,108 @@
+    bucket.delete()
+
+
+@skip("too slow")  # Multipart copy is implicitly tested in import export tests.
```
Speed it up then. Make the part size 5MiB and omit the 128 core case.

JakeNarkizian replied:
The part size already is 5MiB [see](https://github.com/BD2KGenomics/toil/pull/867/files#diff-cf02284aedf452de3ff629898027b42bR31). Also if you recall we agreed since multipart copies are implicitly tested with 2 threads already that we should skip this during regular builds to save time but leave it in as a stress test with 128 included as an edge case. Why have you changed your mind?

hannes-ucsc commented on src/toil/test/src/multipartTransferTest.py:
```diff
@@ -0,0 +1,108 @@
+    """
+    if keySize < 0 and keySize is not None:
+        raise ValueError("Key size must be greater than zero")
+    s3 = boto.connect_s3()
```
You should always use a connection to a specific region.

hannes-ucsc replied:
… which is us-west-2 in this case. Look at AWS job store for a reference.

hannes-ucsc replied:
Also, as usual, you are leaking the connection.

hannes-ucsc commented on src/toil/test/src/multipartTransferTest.py:
```diff
@@ -0,0 +1,108 @@
+    if keySize < 0 and keySize is not None:
+        raise ValueError("Key size must be greater than zero")
+    s3 = boto.connect_s3()
+    bucket = s3.create_bucket('multipart-transfer-test-%s' % uuid.uuid4())
```
Specify a location constraint to create the bucket in the same region as the one you are connected to. Make sure you understand the subtle difference between S3 location and AS region. See bottom aws/utils.py for utility functions about that.

hannes-ucsc commented on src/toil/test/src/multipartTransferTest.py:
```diff
@@ -0,0 +1,108 @@
+        totalSize = keySize
+        totalParts = (totalSize + partSize - 1) / partSize
+        exceptions = collections.OrderedDict()
+        upload = boto.connect_s3().get_bucket(bucket).initiate_multipart_upload(keyName)
```
and leaking the connection

hannes-ucsc commented on src/toil/jobStores/aws/jobStore.py:
```diff
@@ -1158,3 +1121,84 @@ def toItem(self):
+                    try:
+                        return upload.complete_upload()
+                    finally:
+                        log.info("Completed copy from 's3://%s/%s' to 's3://%s/%s'." %
```
This shouldn't be in a finally or you'll get the the log line on every failed (!) attempt.

Common, Jake, these are stupid beginner mistakes.

hannes-ucsc commented on src/toil/test/src/multipartTransferTest.py:
```diff
@@ -0,0 +1,108 @@
+                    assert len(finishedParts) == totalParts
+            upload.complete_upload()
+        except:
+            upload.cancel_upload()
```
Use bd2k-python-lib's (aka BPL) panic() here. Grep the code for a usage example.

JakeNarkizian commented:
@hannes-ucsc I've addressed all your comments except one above which needs clarification, and the two pertaining to the removal of the 128 cores case from multipart transfer test. For reference here is a screenshot of what is now hidden under outdated diffs:
<img width="770" alt="screen shot 2016-06-16 at 9 25 43 pm" src="https://cloud.githubusercontent.com/assets/10471159/16140752/debd4acc-3408-11e6-8781-a7aee6ee43a7.png">

The "see" url links [here](https://github.com/BD2KGenomics/toil/pull/867/files#diff-cf02284aedf452de3ff629898027b42bR31). Not exactly the same place but semantically equivalent.

hannes-ucsc commented on src/toil/test/src/multipartTransferTest.py:
```diff
@@ -0,0 +1,108 @@
+            upload = s3.get_bucket(bucket).initiate_multipart_upload(keyName)
+            try:
+                with open('/dev/urandom', 'rb') as f:
+                    with ThreadPoolExecutor(max_workers=min(cpu_count()*16, totalParts, 128)) as ex:
```
Please remove the thread pool and make this a single-request upload.

hannes-ucsc commented on src/toil/test/src/multipartTransferTest.py:
```diff
@@ -0,0 +1,108 @@
+                                     dstBucket.get_key("test").get_contents_as_string())
+
+        make_tests(_multipartCopy, targetClass=AWSMultipartCopyTest,
+                   threadPoolSize={str(x): x for x in (1, 2, 16, 128)})
```
Remove 128 from the threadPoolSize dimension.

hannes-ucsc commented on src/toil/test/src/multipartTransferTest.py:
```diff
@@ -0,0 +1,108 @@
+        bucket.delete()
+
+
+@skip("too slow")  # Multipart copy is implicitly tested in import export tests.
```
Always include this test.

hannes-ucsc commented on src/toil/test/src/multipartTransferTest.py:
```diff
@@ -0,0 +1,108 @@
+    """
+    if keySize < 0 and keySize is not None:
+        raise ValueError("Key size must be greater than zero")
+    with closing(boto.connect_s3(region='us-west-2')) as s3:
```
Extract 'us-west-2'` into a class attribute.

hannes-ucsc commented on src/toil/test/src/multipartTransferTest.py:
```diff
@@ -0,0 +1,108 @@
+
+    @classmethod
+    def makeTests(self):
+        def _multipartCopy(self, threadPoolSize):
```
Remove underscore. That convention only applies to methods.

hannes-ucsc commented on src/toil/test/src/multipartTransferTest.py:
```diff
@@ -0,0 +1,108 @@
+        def _multipartCopy(self, threadPoolSize):
+            with openS3(threadPoolSize * partSize) as srcKey:
+                with openS3() as dstBucket:
+                    copyKeyMultipart(srcKey, dstBucket, "test")
```
Make all strings single-quoted.

hannes-ucsc replied:
Unless they contain single quotes.

hannes-ucsc commented on src/toil/test/src/multipartTransferTest.py:
```diff
@@ -0,0 +1,108 @@
+            with openS3(threadPoolSize * partSize) as srcKey:
+                with openS3() as dstBucket:
+                    copyKeyMultipart(srcKey, dstBucket, "test")
+                    self.assertEqual(srcKey.get_contents_as_string(),
```
This line here makes the complicated multi-threaded upload in openS3() look ridiculous.

hannes-ucsc commented on src/toil/test/src/multipartTransferTest.py:
```diff
@@ -0,0 +1,108 @@
+                with openS3() as dstBucket:
+                    copyKeyMultipart(srcKey, dstBucket, "test")
+                    self.assertEqual(srcKey.get_contents_as_string(),
+                                     dstBucket.get_key("test").get_contents_as_string())
```
This line here makes the complicated multi-threaded upload in openS3() look ridiculous.

hannes-ucsc commented on src/toil/test/src/multipartTransferTest.py:
```diff
@@ -0,0 +1,108 @@
+    @classmethod
+    def makeTests(self):
+        def _multipartCopy(self, threadPoolSize):
+            with openS3(threadPoolSize * partSize) as srcKey:
```
Pad the size a bit so some threads get used more than once.

hannes-ucsc commented on src/toil/jobStores/aws/jobStore.py:
```diff
@@ -74,38 +82,70 @@ def copyKeyMultipart(srcKey, dstBucketName, dstKeyName, headers=None):
     :rtype: boto.s3.multipart.CompletedMultiPartUpload
     :return: An object representing the completed upload.
     """
+    def _copyPart(partIndex):
```
No underscore.

hannes-ucsc commented on src/toil/test/src/multipartTransferTest.py:
```diff
@@ -0,0 +1,108 @@
+    with closing(boto.connect_s3(region='us-west-2')) as s3:
+        bucket = s3.create_bucket('multipart-transfer-test-%s' % uuid.uuid4(),
+                                  location=region_to_bucket_location('us-west-2'))
+        keyName = uuid.uuid4()
```
This makes keyName not a string.

hannes-ucsc replied:
Which is why the tests should fail. But they won't fail since they are skipped. Funny how it goes, right? And it should explain why I either want this code covered or removed.

hannes-ucsc commented on src/toil/test/src/multipartTransferTest.py:
```diff
@@ -0,0 +1,108 @@
+                    return None
+                else:
+                    try:
+                        part = io.BytesIO(f.read(partSize))
```
You are sharing a file handle between threads which I don't think is ok. But this code is going to go.

JakeNarkizian commented:
@hannes-ucsc

tetron commented:
New failure:

```
ip-172-31-22-175: 2016-06-21 13:51:35,007 INFO: toil.lib.bioio: Logging set at level: DEBUG
ip-172-31-22-175: 2016-06-21 13:51:35,072 DEBUG: toil.jobStores.aws.jobStore: Instantiating <class 'toil.jobStores.aws.jobStore.AWSJobStore'> for region us-west-2 and name prefix 'cwltoil-test-pa2'
ip-172-31-22-175: 2016-06-21 13:51:35,078 INFO: toil.jobStores.aws.jobStore: Setting up job store SDB domain 'toil-registry'.
ip-172-31-22-175: 2016-06-21 13:51:35,123 INFO: toil.jobStores.aws.jobStore: Setting up job store SDB domain 'cwltoil-test-pa2--jobs'.
ip-172-31-22-175: 2016-06-21 13:51:36,392 INFO: toil.jobStores.aws.jobStore: Setting up job store SDB domain 'cwltoil-test-pa2--files'.
ip-172-31-22-175: 2016-06-21 13:51:37,628 INFO: toil.jobStores.aws.jobStore: Setting up job store bucket 'cwltoil-test-pa2--files'.
ip-172-31-22-175: 2016-06-21 13:51:37,628 DEBUG: toil.jobStores.aws.jobStore: Looking up job store bucket 'cwltoil-test-pa2--files'.
ip-172-31-22-175: 2016-06-21 13:51:37,652 DEBUG: toil.jobStores.aws.jobStore: Bucket 'cwltoil-test-pa2--files' does not exist. Creating it.
ip-172-31-22-175: 2016-06-21 13:51:38,535 DEBUG: toil.jobStores.aws.jobStore: Created new job store bucket 'cwltoil-test-pa2--files'.
ip-172-31-22-175: 2016-06-21 13:51:38,569 INFO: toil.jobStores.abstractJobStore: The workflow ID is: 'dfcac354-63b1-4eda-84b3-9b2abdc09c31'
ip-172-31-22-175: 2016-06-21 13:51:38,927 DEBUG: toil.jobStores.aws.jobStore: Wrote FileInfo(fileID='a77581f6-ad82-5142-acb7-5a6b01eab3fb', ownerID='891f7db6-e4d9-4221-a58e-ab6cc4395f94', encrypted=False, version='', previousVersion='', content='\x80\x02ctoil.co...ryCountq:K\x00ub.', _numContentChunks=2) bound to <toil.jobStores.aws.jobStore.AWSJobStore object at 0x7f6c874fc190> for shared file 'config.pickle'.
ip-172-31-22-175: 2016-06-21 13:51:39,014 INFO: toil.jobStores.abstractJobStore: Unable to import 'toil.jobStores.azureJobStore'. This may be because Toil was not installed with the relevant extra.
ip-172-31-22-175: 2016-06-21 13:51:39,411 ERROR: root: Got exception 'S3ResponseError: 403 Forbidden
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>AccessDenied</Code><Message>Access Denied</Message><RequestId>A4AEFEA4FF9FD797</RequestId><HostId>5um1dBgA6B9DIgQvBlhJKkFGeMhCQ5OLdoLEOfgLrclnnMNGyRhZqCk8gQNv3xaKSJ7NNOaAlZA=</HostId></Error>' while writing 's3://1000genomes/phase3/data/HG01977/sequence_read/SRR360135_1.filt.fastq.gz'
Traceback (most recent call last):
  File "/usr/local/bin/cwltoil", line 9, in <module>
    load_entry_point('toil==3.2.0a2', 'console_scripts', 'cwltoil')()
  File "/usr/local/lib/python2.7/dist-packages/toil/cwl/cwltoil.py", line 592, in main
    adjustFiles(builder.job, functools.partial(writeFile, toil.importFile, {}))
  File "/usr/local/lib/python2.7/dist-packages/cwltool/process.py", line 128, in adjustFiles
    adjustFiles(rec[d], op)
  File "/usr/local/lib/python2.7/dist-packages/cwltool/process.py", line 131, in adjustFiles
    adjustFiles(d, op)
  File "/usr/local/lib/python2.7/dist-packages/cwltool/process.py", line 131, in adjustFiles
    adjustFiles(d, op)
  File "/usr/local/lib/python2.7/dist-packages/cwltool/process.py", line 126, in adjustFiles
    rec["path"] = op(rec["path"])
  File "/usr/local/lib/python2.7/dist-packages/toil/cwl/cwltoil.py", line 162, in writeFile
    index[x] = (writeFunc(rp), os.path.basename(x))
  File "/usr/local/lib/python2.7/dist-packages/toil/common.py", line 677, in importFile
    return self._jobStore.importFile(srcUrl, sharedFileName=sharedFileName)
  File "/usr/local/lib/python2.7/dist-packages/toil/jobStores/abstractJobStore.py", line 271, in importFile
    return self._importFile(otherCls, url, sharedFileName=sharedFileName)
  File "/usr/local/lib/python2.7/dist-packages/toil/jobStores/aws/jobStore.py", line 335, in _importFile
    srcKey = self._getKeyForUrl(url, existing=True)
  File "/usr/local/lib/python2.7/dist-packages/toil/jobStores/aws/jobStore.py", line 395, in _getKeyForUrl
    region = bucket_location_to_region(s3.get_bucket(url.netloc).get_location())
  File "/usr/local/lib/python2.7/dist-packages/boto/s3/bucket.py", line 1144, in get_location
    response.status, response.reason, body)
boto.exception.S3ResponseError: S3ResponseError: 403 Forbidden
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>AccessDenied</Code><Message>Access Denied</Message><RequestId>A4AEFEA4FF9FD797</RequestId><HostId>5um1dBgA6B9DIgQvBlhJKkFGeMhCQ5OLdoLEOfgLrclnnMNGyRhZqCk8gQNv3xaKSJ7NNOaAlZA=</HostId></Error>
```

tetron commented:
The error message suggests that it is trying to write to `s3://1000genomes/phase3/data/HG01977/sequence_read/SRR360135_1.filt.fastq.gz` instead of read from it ?????

hannes-ucsc commented:
The "error while writing" log entry is made by cwltoil.py. I don't know why it says "writing".

Anyways, it's `[GetBucketLocation](http://docs.aws.amazon.com/AmazonS3/latest/API/RESTBucketGETlocation.html)` that fails. I'm assuming you don't control permissions on that bucket, correct? Any S3 bucket policy that allows read only access should also allow GetBucketLocation. We might be able to work around this by catching the exception and using the generic endpoint.

hannes-ucsc commented:
> That's a valid question. I don't know, I will check.

I'm assuming the key was indeed missing. Let us know if that's not the case.

hannes-ucsc commented on src/toil/test/src/multipartTransferTest.py:
```diff
@@ -0,0 +1,88 @@
+    def makeTests(self):
+        def multipartCopy(self, threadPoolSize):
+            # key size is padded to ensure some threads are reused
+            keySize = int(math.ceil((threadPoolSize * partSize) * 1.3))
```
ceil is redundant

hannes-ucsc commented on src/toil/test/src/multipartTransferTest.py:
```diff
@@ -0,0 +1,88 @@
+            key.set_contents_from_string(content)
+
+            yield bucket.get_key(keyName)
+        for key in bucket.list():
```
these need to be in a `finally:`

hannes-ucsc replied:
You need to burn the following pattern into your brain:

```
allocateResource()
try:
    useResource()
finally:
   freeResource()
```

hannes-ucsc commented on src/toil/test/src/multipartTransferTest.py:
```diff
@@ -0,0 +1,88 @@
+        bucket = s3.create_bucket('multipart-transfer-test-%s' % uuid.uuid4(),
+                                  location=region_to_bucket_location('us-west-2'))
+        keyName = 'test'
+        keySize = keySize or -1
```
why? just test for None in the `if` below

JakeNarkizian replied:
I was trying to eliminate the pycharm warning on line 59 for passing a possible None instead of an int to os.urandom. But its inconsequential since None will never reach that line of code anyway.

hannes-ucsc commented on src/toil/test/src/multipartTransferTest.py:
```diff
@@ -0,0 +1,88 @@
+
+    :param int keySize: Size of key to be created.
+    """
+    if keySize < 0 and keySize is not None:
```
Due to a Python quirk `None` is less than `0` but I wouldn't rely on this and swap the LHS and RHS of the `and`

hannes-ucsc commented on src/toil/test/src/multipartTransferTest.py:
```diff
@@ -0,0 +1,88 @@
+@contextmanager
+def openS3(keySize=None):
+    """
+    Creates an AWS bucket. If keySize is given a key of random bytes is created and its handle
```
comma after `given`

hannes-ucsc commented on src/toil/test/src/multipartTransferTest.py:
```diff
@@ -0,0 +1,88 @@
+        raise ValueError('Key size must be greater than zero')
+    with closing(boto.s3.connect_to_region(AWSMultipartCopyTest.region)) as s3:
+        bucket = s3.create_bucket('multipart-transfer-test-%s' % uuid.uuid4(),
+                                  location=region_to_bucket_location('us-west-2'))
```
I knew your we're going to miss this one. Use `AWSMultipartCopyTest.region` here as well.

hannes-ucsc commented on src/toil/jobStores/aws/jobStore.py:
```diff
@@ -74,38 +82,70 @@ def copyKeyMultipart(srcKey, dstBucketName, dstKeyName, headers=None):
+                log.debug('Successfully copied part %d of %d' % (partIndex, totalParts))
+
+            except Exception as e:
+                if partIndex <= 4:
```
Not the first four parts of the key. The first four failures. With this code, if the first four parts succeed but another part fails, weird things are going to happen.

hannes-ucsc commented on src/toil/jobStores/aws/jobStore.py:
```diff
@@ -74,38 +82,70 @@ def copyKeyMultipart(srcKey, dstBucketName, dstKeyName, headers=None):
-                                                  headers=headers)
-                start += partSize
+            with ThreadPoolExecutor(max_workers=min(cpu_count() * 16, totalParts, 128)) as executor:
+                finishedParts = set(executor.map(copyPart, range(0, totalParts)))
```
xrange() is probably better in this case

hannes-ucsc commented on src/toil/jobStores/aws/jobStore.py:
```diff
@@ -74,38 +82,70 @@ def copyKeyMultipart(srcKey, dstBucketName, dstKeyName, headers=None):
+                finishedParts = set(executor.map(copyPart, range(0, totalParts)))
+                if exceptions:
+                    for partIndex, tb in exceptions[0:4].items():
+                        log.info("Part %d of copy failed with the following exception:\n%s" %
```
https://github.com/BD2KGenomics/toil/wiki/Python-Dos-and-Don'ts#dont-use-format-or--to-format-log-messages

hannes-ucsc commented on src/toil/jobStores/aws/jobStore.py:
```diff
@@ -74,38 +82,70 @@ def copyKeyMultipart(srcKey, dstBucketName, dstKeyName, headers=None):
                 with attempt:
-                    return upload.complete_upload()
+                    completed = upload.complete_upload()
+                    log.info("Completed copy from 's3://%s/%s' to 's3://%s/%s'." %
```
https://github.com/BD2KGenomics/toil/wiki/Python-Dos-and-Don'ts#dont-use-format-or--to-format-log-messages

hannes-ucsc commented on src/toil/jobStores/aws/jobStore.py:
```diff
@@ -357,20 +397,24 @@ def _getKeyForUrl(url, existing=None):
             if existing is True:
                 if key is None:
-                    raise RuntimeError('Key does not exist.')
+                    raise RuntimeError("Key '%s' does not exist in bucket '%s'." %
```
https://github.com/BD2KGenomics/toil/wiki/Python-Dos-and-Don'ts#dont-use-format-or--to-format-log-messages

hannes-ucsc replied:
Never mind.

hannes-ucsc commented on src/toil/jobStores/aws/jobStore.py:
```diff
@@ -74,38 +82,70 @@ def copyKeyMultipart(srcKey, dstBucketName, dstKeyName, headers=None):
+                                                         end=min(partIndex * partSize + partSize,
+                                                                 totalSize) - 1,
+                                                         headers=headers)
+                log.debug('Successfully copied part %d of %d' % (partIndex, totalParts))
```
https://github.com/BD2KGenomics/toil/wiki/Python-Dos-and-Don'ts#dont-use-format-or--to-format-log-messages

hannes-ucsc commented on src/toil/jobStores/aws/jobStore.py:
```diff
@@ -74,38 +82,70 @@ def copyKeyMultipart(srcKey, dstBucketName, dstKeyName, headers=None):
                 dstBucket = s3.get_bucket(dstBucketName)
                 upload = dstBucket.initiate_multipart_upload(dstKeyName, headers=headers)
+        log.info("Initiated multipart copy from 's3://%s/%s' to 's3://%s/%s'." %
+                 (srcKey.bucket.name, srcKey.name, dstBucketName, dstKeyName))
```
https://github.com/BD2KGenomics/toil/wiki/Python-Dos-and-Don'ts#dont-use-format-or--to-format-log-messages

hannes-ucsc commented:
@JakeNarkizian, done with review.

tetron commented:
@hannes-ucsc correct on both counts.  I checked and it looks the file is indeed missing, but it looks like I may be able to remove it as a workflow input.  For the confusing error message from cwltoil, I will do a PR to improve the error message (it should say "error X while copying Y to the file store"), thanks for pointing that out.

I don't know why the public 1000 genomes bucket would return a 403 error, but you are correct that I don't have any control over it.

hannes-ucsc commented:
Alright, so Jake is going to add a workaround for this then.

JakeNarkizian commented:
Previous test failure was mostly due to the number of domains on sub being exceeded. A couple were due to a misplaced panic() in multipartTransferTest. This should be fixed now.

hannes-ucsc commented:
Jenkins, test this please.

hannes-ucsc commented on src/toil/jobStores/aws/jobStore.py:
```diff
@@ -351,26 +391,36 @@ def _getKeyForUrl(url, existing=None):
+                raise
         # Note that caller is responsible for closing the connection
-        s3 = boto.s3.connect_to_region(region)
+        s3 = boto.s3.connect_to_region(region) or boto.connect_s3()
```
This is relying on an implementation detail. The fact that connect_to_region() gracefully handles None is undocumented.You know off-hand which call to make so why even try connect_to_region().

hannes-ucsc replied:
… if region is None.

hannes-ucsc commented:
You should also write a test that covers the 403 workaround for get_location(). For now, lets just the URL that caused the problem for @tetron.

hannes-ucsc commented on src/toil/jobStores/aws/jobStore.py:
```diff
@@ -349,28 +389,54 @@ def _getKeyForUrl(url, existing=None):
                exists and it will be created. If None, the key will be created if it doesn't exist.
 
         :rtype: Key
+        >>> import urlparse
```
Why all this duplication?

JakeNarkizian replied:
I was trying to create the below code so that it can be covered in a test. I think I might have misunderstood what you had in mind for this test.

hannes-ucsc replied:
I changed my mind: no doctest, please write a regular unit test without duplicating any code.

JakeNarkizian commented:
Error tracked in #995

JakeNarkizian commented:
Jenkins, test this please.

hannes-ucsc commented:
I'll take it from here.

hannes-ucsc commented on src/toil/jobStores/aws/jobStore.py:
```diff
@@ -74,38 +82,70 @@ def copyKeyMultipart(srcKey, dstBucketName, dstKeyName, headers=None):
+            with ThreadPoolExecutor(max_workers=min(cpu_count() * 16, totalParts, 128)) as executor:
+                finishedParts = set(executor.map(copyPart, xrange(0, totalParts)))
+                if exceptions:
+                    for partIndex, tb in exceptions[0:4].items():
```
You do not need to act on this, but I want you to know that using this slice here is still incorrect. The lesson we should learn is that we need better test coverage to expose these issues. Please study my upcoming commit to see how I dealt with it.

hannes-ucsc commented:
@JakeNarkizian, I encourage you to review cc5fe9eb825be658892cd8a0e379874f3856f38d.
